src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/Diagnostics/RemoteDiagnosticsService.cs

```diff
@@ -0,0 +1,59 @@
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.CodeAnalysis.ExternalAccess.Razor;
+using Microsoft.CodeAnalysis.Razor.Diagnostics;
+using Microsoft.CodeAnalysis.Razor.Protocol;
+using Microsoft.CodeAnalysis.Razor.Remote;
+using Microsoft.CodeAnalysis.Remote.Razor.ProjectSystem;
+using LspDiagnostic = Microsoft.VisualStudio.LanguageServer.Protocol.Diagnostic;
+
+namespace Microsoft.CodeAnalysis.Remote.Razor;
+
+internal sealed class RemoteDiagnosticsService(in ServiceArgs args) : RazorDocumentServiceBase(in args), IRemoteDiagnosticsService
+{
+    internal sealed class Factory : FactoryBase<IRemoteDiagnosticsService>
+    {
+        protected override IRemoteDiagnosticsService CreateService(in ServiceArgs args)
+            => new RemoteDiagnosticsService(in args);
+    }
+
+    private readonly RazorTranslateDiagnosticsService _translateDiagnosticsService = args.ExportProvider.GetExportedValue<RazorTranslateDiagnosticsService>();
+
+    public ValueTask<ImmutableArray<LspDiagnostic>> GetDiagnosticsAsync(
+        JsonSerializableRazorPinnedSolutionInfoWrapper solutionInfo,
+        JsonSerializableDocumentId documentId,
+        LspDiagnostic[] csharpDiagnostics,
+        LspDiagnostic[] htmlDiagnostics,
+        CancellationToken cancellationToken)
+        => RunServiceAsync(
+            solutionInfo,
+            documentId,
+            context => GetDiagnosticsAsync(context, csharpDiagnostics, htmlDiagnostics, cancellationToken),
+            cancellationToken);
+
+    private async ValueTask<ImmutableArray<LspDiagnostic>> GetDiagnosticsAsync(
+        RemoteDocumentContext context,
+        LspDiagnostic[] csharpDiagnostics,
+        LspDiagnostic[] htmlDiagnostics,
+        CancellationToken cancellationToken)
+    {
+        // We've got C# and Html, lets get Razor diagnostics
+        var codeDocument = await context.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+        // Yes, CSharpDocument.Documents are the Razor diagnostics. Don't ask.
+        var razorDiagnostics = codeDocument.GetCSharpDocument().Diagnostics;
+
+        using var allDiagnostics = new PooledArrayBuilder<LspDiagnostic>(capacity: razorDiagnostics.Length + csharpDiagnostics.Length + htmlDiagnostics.Length);
+
+        allDiagnostics.AddRange(RazorDiagnosticConverter.Convert(razorDiagnostics, codeDocument.Source.Text, context.Snapshot));
+        allDiagnostics.AddRange(await _translateDiagnosticsService.TranslateAsync(RazorLanguageKind.CSharp, csharpDiagnostics, context.Snapshot));
+        allDiagnostics.AddRange(await _translateDiagnosticsService.TranslateAsync(RazorLanguageKind.Html, htmlDiagnostics, context.Snapshot));
+
+        return allDiagnostics.DrainToImmutable();
```

Comment:
Since this code is just joining arrays, another option might be to avoid the builder completely and use a collection expression.

```suggestion
        return [
          .. RazorDiagnosticConverter.Convert(razorDiagnostics, codeDocument.Source.Text, context.Snapshot),
          .. await _translateDiagnosticsService.TranslateAsync(RazorLanguageKind.CSharp, csharpDiagnostics, context.Snapshot),
          .. await _translateDiagnosticsService.TranslateAsync(RazorLanguageKind.Html, htmlDiagnostics, context.Snapshot)
        ];
```

This is totally a suggested possibility though. I just wanted to point out that the code produced by a collection expression is very tight in this sort of scenario. It just creates an array of the correct length and each array's data. [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgPgAQAwAIEEYB0BhA9gG3wFMBjAFylxgGdMBJMMAVzImGIG4BYAKGTSwAZWAEdufVBgCs43ggDMaAEwpsKAN68U2tIoQAOFADUI+JkQQA2ADy5gAK1JkA2gF0AfCgDiRMgBEoCABzGFxqChJqAEFqAE8YEgAKLR1UhmZWdiIogCcciFjrDCRPfIAvXByA4NDwqEiAGhQ7R3I3FEiACwgcgAdqkLCI6iaWp3bOsjB8AdrhgEoNFNTtZZWEAHYUZzWVlExMVSoANyIcskTyytmh+up5ht2Vg7QAThQAFXyafAgybLiCUSXR6/UCgzqkQeT1SLwQ7y+EB+fwB8SSk2mN0h9xhKFc4lSAF81msxm1XEcYKdzol0iw2MRcvlCsVPAATcFzO7zNaaHh7bSbFAcmq3SKYD64JkFRLzAk6Yn8nRrKy2BzjDyfb7UX7/GJoxJklwUkUQha83FCqyYABiOVwYAASkRqEx8GQ1a1je5EqauVD5dpFYSgA===)